File: solution/0100-0199/0132.Palindrome Partitioning II/Solution.cs
public class Solution {
    public int MinCut(string s) {
        int n = s.Length;
        bool[,] g = new bool[n,n];
        int[] f = new int[n];
        for (int i = 0; i < n; ++i) {
            f[i] = i;
            for (int j = 0; j < n; ++j) {
                g[i,j] = true;
            }
        }
        for (int i = n - 1; i >= 0; --i) {
            for (int j = i + 1; j < n; ++j) {
                g[i,j] = s[i] == s[j] && g[i + 1,j - 1];
            }
        }
        for (int i = 1; i < n; ++i) {
            for (int j = 0; j <= i; ++j) {
                if (g[j,i]) {
                    f[i] = Math.Min(f[i], j > 0 ? 1 + f[j - 1] : 0);
                }
            }
        }
        return f[n - 1];
    }
}

File: solution/0100-0199/0132.Palindrome Partitioning II/Solution.java
class Solution {
    public int minCut(String s) {
        if(s==null || s.length()<=1)return 0;
        int len = s.length();
        int[] dp = new int[len];
        for(int i=0;i<len;i++) dp[i] = len - 1;
        for(int i=0;i<len;i++){
            mincutHelper(s , i , i , dp);
            mincutHelper(s, i , i+1 , dp);
        }
        return dp[len-1];
    }
    private void mincutHelper(String s, int i, int j, int[] dp){
        int len = s.length();
        while(i>=0 && j<len && s.charAt(i)==s.charAt(j)){
            dp[j] = Math.min(dp[j] , (i==0?-1:dp[i-1])+1);
            i--;
            j++;
        }
    }
}